Soundex.cs

```diff
@@ -1,68 +1,68 @@
 {
-    public static string GenerateSoundex(string name)
+public static string GenerateSoundex(string name)
+{
+    if(string.IsNullOrEmpty(name))
+    {
+        return string.Empty;
+    }
+    StringBuilder soundex = InitializeTheSoundex(name);
+    char prevCode = GetSoundexCode(name[0]);
+ 
+    AppendingSoundexCharacters(name, soundex, ref prevCode);
+    SoundexCode(ref soundex);
+    return soundex.ToString();
+}
+ 
+public static StringBuilder InitializeTheSoundex(string name)
+{
+    StringBuilder soundex = new StringBuilder();
+    soundex.Append(char.ToUpper(name[0]));
+    return soundex;
+}
+ 
+public static void AppendingSoundexCharacters(string name, StringBuilder soundex, ref char prevCode)
+{
+    for (int i = 1; i < name.Length && soundex.Length < 4; i++)
+    {
+        Characters(name[i], soundex, ref prevCode); 
+    }
+}
+public static void Characters(char character, StringBuilder soundex, ref char prevCode)
+{
+   if(char.IsLetter(character))
+    {
+       char code = GetSoundexCode(character);
+    if (AppendCode(code, prevCode))
     {
-        if (string.IsNullOrEmpty(name))
-        {
-            return string.Empty;
-        }
-
-        StringBuilder soundex = new StringBuilder();
-        soundex.Append(char.ToUpper(name[0]));
-        char prevCode = GetSoundexCode(name[0]);
-
-        for (int i = 1; i < name.Length && soundex.Length < 4; i++)
-        {
-            char code = GetSoundexCode(name[i]);
-            if (code != '0' && code != prevCode)
-            {
-                soundex.Append(code);
-                prevCode = code;
-            }
-        }
-
-        while (soundex.Length < 4)
-        {
-            soundex.Append('0');
-        }
-
-        return soundex.ToString();
+        soundex.Append(code);
+        prevCode = code;
+    } 
     }
-
-    private static char GetSoundexCode(char c)
+}
+public static bool AppendCode(char code, char prevCode) => code != '0' && code != prevCode;
+public static void SoundexCode(ref StringBuilder soundex)
+{
+    while (soundex.Length < 4)
     {
-        c = char.ToUpper(c);
-        switch (c)
-        {
-            case 'B':
-            case 'F':
-            case 'P':
-            case 'V':
-                return '1';
-            case 'C':
-            case 'G':
-            case 'J':
-            case 'K':
-            case 'Q':
-            case 'S':
-            case 'X':
-            case 'Z':
-                return '2';
-            case 'D':
-            case 'T':
-                return '3';
-            case 'L':
-                return '4';
-            case 'M':
-            case 'N':
-                return '5';
-            case 'R':
-                return '6';
-            default:
-                return '0'; // For A, E, I, O, U, H, W, Y
-        }
+        soundex.Append('0');
     }
 }
+public static char GetSoundexCode(char character)
+{
+    character = char.ToUpper(character);
+    return character switch
+    {

```

Comment:
how does this code will reduce complexity ?

Reply:
This is handled by replacing it with a dictionary.